### PR TITLE
fix: fix route mismatching

### PR DESCRIPTION
**Pull request by JerryWu1234**
## Changes

Regex sometimes fails to match. I splice the complete path and match it to ensure that it is completely correct.
#6968

### CODE REVIEW

File: .changeset/witty-grapes-jam.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': major
```
Comment (natemoo-re) [inline .changeset/witty-grapes-jam.md:2]: This is a bug fix.

```suggestion
'astro': patch
```

File: packages/astro/src/core/render/route-cache.ts [hunk 2]
```diff
@@ -126,3 +127,47 @@ export function findPathItemByKey(
...
+					}
+				});
+			});
+			if (pathname.endsWith(substr)) {
+				const params = Object.keys(arg.params as Params);
+				// related to routing-priority.test.js /empty-slug/undefined
+				// if the params is undefined, it will be 'undefined'
+				return params.reduce((obj, key: string) => {
+					if (arg.params?.[key] === undefined) {
+						obj[key] = 'undefined';
```
[the remark this answers is not in the document]
Comment (JerryWu1234) [reply packages/astro/src/core/render/route-cache.ts:163]: related to routing-priority.test.js /empty-slug/undefined
because of this

**Comment by JerryWu1234** [comment]
Thank you for reviewing the code. 
I’m a little busy today. 
I will fix it next Monday. Welcome comments.

**Comment by natemoo-re** [comment]
I'm going to close this as https://github.com/withastro/astro/pull/7694 fixed the same issue! Thanks for opening the PR, though!

**Pull request closed without merging**